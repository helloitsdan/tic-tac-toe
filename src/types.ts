export enum BoardPosition {
  TOP_LEFT,
  TOP_CENTRE,
  TOP_RIGHT,
  MIDDLE_LEFT,
  MIDDLE_CENTRE,
  MIDDLE_RIGHT,
  BOTTOM_LEFT,
  BOTTOM_CENTRE,
  BOTTOM_RIGHT,
}

export enum PlayerPiece {
  PLAYER_ONE = 'PLAYER_ONE',
  PLAYER_TWO = 'PLAYER_TWO',
}

export enum MatchName {
  TOP_ROW = 'TOP_ROW',
  MIDDLE_ROW = 'MIDDLE_ROW',
  BOTTOM_ROW = 'BOTTOM_ROW',
  LEFT_COLUMN = 'LEFT_COLUMN',
  CENTRE_COLUMN = 'CENTRE_COLUMN',
  RIGHT_COLUMN = 'RIGHT_COLUMN',
  DIAGONAL_LEFT_TO_RIGHT = 'DIAGONAL_LEFT_TO_RIGHT',
  DIAGONAL_RIGHT_TO_LEFT = 'DIAGONAL_RIGHT_TO_LEFT',
}

export interface PlayerType {
  piece: PlayerPiece
  name: string
}

export type MatchTuple = [MatchName, BoardPosition[]]

export interface PlayerMatchType {
  match: MatchTuple
  piece: PlayerPiece
}

export type GameMovesType = Record<PlayerPiece, BoardPosition[]>

export interface GameStateType {
  players: PlayerType[]
  moves: GameMovesType
  winner: PlayerMatchType
}
