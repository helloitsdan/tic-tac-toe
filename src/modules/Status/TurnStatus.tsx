import React, { FunctionComponent, useMemo } from 'react'
import { useSelector } from 'react-redux'
import { getPlayers } from '../../store/selectors'

const TurnStatus: FunctionComponent = () => {
  const { active } = useSelector(getPlayers)
  const players = useSelector(getPlayers)

  const player = useMemo(() => players.all.find(p => p.piece === active), [active, players])

  return (
    <div data-testid="TurnStatus">
      {player ? (
        <>
          <span data-testid="TurnStatus/Player">{player.name}</span>'s turn!
        </>
      ) : (
        <span data-testid="TurnStatus/UnknownPlayer">Your turn!</span>
      )}
    </div>
  )
}

export default TurnStatus
