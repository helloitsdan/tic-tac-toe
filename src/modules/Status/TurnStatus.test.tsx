import React from 'react'
import { render } from '@testing-library/react'

import TurnStatus from './TurnStatus'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { PlayerId } from '../../types'

describe('TurnStatus', () => {
  const testPlayers = {
    active: PlayerId.PLAYER_ONE,
    all: [
      { piece: PlayerId.PLAYER_ONE, name: 'Test Player One' },
      { piece: PlayerId.PLAYER_TWO, name: 'Test Player Two' },
    ]
  }

  it('renders', () => {
    const testStore = createStore(() => ({
      players: testPlayers,
    }))

    const { asFragment } = render(
      <Provider store={testStore}>
        <TurnStatus />
      </Provider>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders the name of the currently active player', () => {
    const testStore = createStore(() => ({
      players: testPlayers,
    }))

    const { getByText } = render(
      <Provider store={testStore}>
        <TurnStatus />
      </Provider>
    )
    
    expect(getByText(testPlayers.all[0].name)).not.toBeNull()
  })

  it('handles unknown players', () => {
    const testStore = createStore(() => ({
      players: {
        ...testPlayers,
        active: 'A_MYSTERY_PERSON',
      },
    }))

    const { getByTestId } = render(
      <Provider store={testStore}>
        <TurnStatus />
      </Provider>
    )
    
    expect(getByTestId('TurnStatus/UnknownPlayer')).not.toBeNull()
  })
})
