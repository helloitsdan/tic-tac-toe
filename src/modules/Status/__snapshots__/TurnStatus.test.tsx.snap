// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TurnStatus renders 1`] = `
<DocumentFragment>
  <div
    data-testid="TurnStatus"
  >
    <span
      data-testid="TurnStatus/Player"
    >
      Test Player One
    </span>
    's turn!
  </div>
</DocumentFragment>
`;
